source/store_test.go

```diff
@@ -17,86 +17,73 @@ limitations under the License.
 package source
 
 import (
+	"errors"
 	"testing"
 
-	"github.com/kubernetes-incubator/external-dns/internal/testutils"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
 
-	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
 )
 
-func TestStore(t *testing.T) {
-	t.Run("RegisterAndLookup", testRegisterAndLookup)
-	t.Run("LookupMultiple", testLookupMultiple)
+type MockClientProvider struct {
+	mock.Mock
+	client kubernetes.Interface
 }
 
-// testRegisterAndLookup tests that a Source can be registered and looked up by name.
-func testRegisterAndLookup(t *testing.T) {
-	for _, tc := range []struct {
-		title            string
-		givenAndExpected map[string]Source
-	}{
-		{
-			"registered source is found by name",
-			map[string]Source{
-				"foo": &testutils.MockSource{},
-			},
-		},
-	} {
-		t.Run(tc.title, func(t *testing.T) {
-			for k, v := range tc.givenAndExpected {
-				Register(k, v)
-			}
-
-			for k, v := range tc.givenAndExpected {
-				assert.Equal(t, v, Lookup(k))
-			}
-		})
+func (m *MockClientProvider) KubeClient() (kubernetes.Interface, error) {
+	args := m.Called()
+	if args.Error(1) == nil {
+		m.client = args.Get(0).(kubernetes.Interface)
+		return m.client, nil
 	}
+	return nil, args.Error(1)
 }
 
-// testLookupMultiple tests that Sources can be looked up by providing multiple names.
-func testLookupMultiple(t *testing.T) {
-	for _, tc := range []struct {
-		title       string
-		registered  map[string]Source
-		names       []string
-		expectError bool
-	}{
-		{
-			"multiple registered sources are found by names",
-			map[string]Source{
-				"foo": &testutils.MockSource{},
-				"bar": &testutils.MockSource{},
-			},
-			[]string{"foo", "bar"},
-			false,
-		},
-		{
-			"multiple registered sources, one source not registered",
-			map[string]Source{
-				"foo": &testutils.MockSource{},
-				"bar": &testutils.MockSource{},
-			},
-			[]string{"foo", "bar", "baz"},
-			true,
-		},
-	} {
-		t.Run(tc.title, func(t *testing.T) {
-			for k, v := range tc.registered {
-				Register(k, v)
-			}
-
-			lookup, err := LookupMultiple(tc.names)
-			if tc.expectError {
-				require.Error(t, err)
-			} else {
-				require.NoError(t, err)
-				require.Len(t, lookup, len(tc.registered))
-				for _, source := range tc.registered {
-					assert.Contains(t, lookup, source)
-				}
-			}
-		})
-	}
+type ByNamesTestSuite struct {
+	suite.Suite
+}
+
+func (suite *ByNamesTestSuite) TestAllInitialized() {
+	mockClientProvider := new(MockClientProvider)
+	mockClientProvider.On("KubeClient").Return(fake.NewSimpleClientset(), nil)
```

Comment:
Instead of mocking we could also give `ClientProvider` itself the ability to be configured with a client, e.g.

```go
p := NewClientProvider(x)
p.Client() => x
```

Seems silly but `ClientProvider` is supposed to be like a "container" for a kubernetes client, right?